 syntax = "proto3";

package memmgr;

service MemoryManager{
     rpc Create (CreateRequest) returns (CreateResponse);
     rpc Set (SetRequest) returns (SetResponse);
     rpc Get (GetRequest) returns (GetResponse);
     rpc IncreaseRefCount (IncreaseRefCountRequest) returns (IncreaseRefCountResponse);
     rpc DecreaseRefCount (DecreaseRefCountRequest) returns (DecreaseRefCountResponse);
}

message CreateRequest{
  uint32 size = 1;
  string type = 2;
}

message CreateResponse{
  uint64 id = 1;
}

message SetRequest{
  uint64 id = 1;
  bytes value = 2;
}

message SetResponse{
  bool success = 1;
}

 message GetRequest{
   uint64 id = 1;
 }

 message GetResponse{
   bytes value = 1;
 }

 message IncreaseRefCountRequest{
  uint64 id = 1;
 }

 message IncreaseRefCountResponse{
   bool success = 1;
 }

 message DecreaseRefCountRequest{
   uint64 id = 1;
 }

 message DecreaseRefCountResponse{
   bool success = 1;
 }

